21.md

```diff
@@ -0,0 +1,209 @@
+  multiplier: number, // Number of millisatoshis per smallets unit of currency. E.g.: 5405.405
+  convertible?: bool // Whether the currency can be converted into.
+}
+
+type ExtendedResponse = BaseResponse & {
+  currencies: Currency[]
+}
+```
+
+```diff
+{
+  "tag": "payRequest",
+  "metadata": '[["text/plain","$kenu ⚡ bipa.app"]]',
+  "callback": "https://api.bipa.app/ln/request/invoice/kenu",
+  "maxSendable": 1000000000,
+  "minSendable": 1000,
++ "currencies": [
++   {
++     "code": "BRL",
++     "name": "Real",
++     "symbol": "R$",
++     "decimals": 2,
++     "multiplier": 5404.405,
++     "convertible": true
++   }
++ ]
+}
+```
+
+- The inclusion of the `currencies` field implies the support of this extension
+- The inclusion of a `currency` implies it can be used for denomination of an amount
+- The inclusion of a `convertible currency` implies the `SERVICE` can quote and guarantee a price for given currency
+- The price returned on this response is not guaranteed by the `SERVICE` and is subject to change
+- The `code` of a `currency` will be used as an identifier for the next request
+- The `code` and other information must be according to [ISO-4217](https://en.wikipedia.org/wiki/ISO_4217) for currencies that are included in it
+- The order of the `currencies` must be interpreted by the `WALLET` as the receiving user preference for a currency
+
+### Wallet-side second request
+
+Upon seeing the `currencies` field on the response from `SERVICE`, a `WALLET` may show the user it has the option of denominating the amount in different currencies or even of sending sats to be credited as the different currency for the receiver.
+
+After gathering the input from the user, it must be encoded into the callback as query parameters as follows.
+
+The most general case, when the user is denominating the amount in `CURRENCY1`, but wants it to be converted to `CURRENCY2` on `SERVICE`.
+`CURRENCY1` and `CURRENCY2` can be the same, which would the common use case for remittances.
+```diff
+- <callback><?|&>amount=<milliSatoshi>
++ <callback><?|&>amount=<AMOUNT><CURRENCY1>&convert=<CURRENCY2>
+
+In a more particular case, when user is denominating the amount in `CURRENCY`, but does not want to convert it or the service does not support conversion.
+```diff
+- <callback><?|&>amount=<milliSatoshi>
++ <callback><?|&>amount=<AMOUNT><CURRENCY>
+```
+
+In another particular case, when the user is denominating the amount in millisatoshi, but wants it to be converted to `CURRENCY` on `SERVICE`.
+```diff
+- <callback><?|&>amount=<millisatoshi>
++ <callback><?|&>amount=<millisatoshi>&convert=<CURRENCY>
+```
+
+Note that the amount provided in all requests is always an integer number denominated in the smallets possible unit of the selected currency. The smallest unit need to be in sync with the `decimals` parameter, so the `WALLET` has all the needed information to receive input and show output properly.
```

Comment:
I just want to say that It's really confusing to request 100USDT, intuitively expecting to be `$ 100.00` not `$ 1.00`.
I have no other suggestions in fixing this for now, just my 002 cents

Reply:
Yeah, I agree this is a weak point of the proposal, but I also didn't come with anything better. Things I thought of were: (1) having the currency before the value like USDT100 but it amounts to the same problem; (2) having the currency on another field like amount=100&currency=USDT but again, it doesn't help much and makes the request ambiguous for non-implementers of the LUD.

Just throwing ideas. I'd love to have a less confusing/ambiguous way of expressing the same idea.

> just my 002 cents
haha got me there

Reply:
If having unambiguous interface isn't a priority, this could be converted to something like `amount={A}&currency={C1}&convert={C2}` or even `amount={A}&input={C1}&convert={C2}`. I gave preference for having the error on non-supporting wallets if you pass the wrong parameters, but it could also be interpreted as client error so we don't need to mix stuff on the amount field.

We discussed this privately too, so feel free to give your 002 cents if you want @jklein24